### PR TITLE
jss transform: transform via postcss

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Transform the extracted css via our custom `postcss` function.

This is pretty gnarly. `postcss` runs async but babel can only run sync. I get around this by using `spawnSync`, but it is needlessly expensive.  When we start transforming more files it may pay to invest in starting up an external server and communicate over ports to amortize the cost as they do in [babel-plugin-transform-css](https://github.com/wbyoung/babel-plugin-transform-postcss).

The more obvious and better solution is to use jss plugins. this seems harebrained

### CODE REVIEW

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js [hunk 1]
```diff
@@ -126,3 +127,18 @@ module.exports = function ({types: t, template}) {
...
   };
 };
+
+// Abuses spawnSync to let us run an async function sync.
+function transformCssSync(cssText) {
+  const programText = `
+  const {transformCss} = require('../../../build-system/tasks/jsify-css');
+  transformCss(\`${cssText}\`).then((css) => console.log(css.toString()));
+  `;
+  const spawnedProcess = spawnSync(`node`, [`-e`, programText], {
```
Comment (dvoytenko) [inline build-system/babel-plugins/babel-plugin-transform-jss/index.js:137]: nit: normal quotes for `-e` and `node`?

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js [hunk 2]
```diff
@@ -126,3 +127,18 @@ module.exports = function ({types: t, template}) {
     },
   };
 };
+
+// Abuses spawnSync to let us run an async function sync.
+function transformCssSync(cssText) {
+  const programText = `
```
Comment (dvoytenko) [inline build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: Can we actually do it like this? Or should we be calling a gulp task? I.e. what if the `jsify-css` starts using some gulp API?

**Review by dvoytenko — APPROVED** [review]

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js (hunk 1 above)
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:137]: sure!

**Comment by samouri** [comment]
After discussing with @dvoytenko: seems like we actually want to pursue this path for now. It keeps us more flexible to pivot away from JSS if we choose, and we get to have the confidence that the same optimizer is used for both bento and pre-bento components.

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js (hunk 2 above)
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: I'm not sure what you mean. If `jsify-css` changes in a way that breaks this function, the tests will fail.

I don't know how we could pass all of this as an arg to a gulp task and whether or not we can get the response back in an sensible way. Justin alerted me that we have some exec helpers in build system I'll check out.

Comment (dvoytenko) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: Basically my question is only: our build system is built on gulp and I'm not sure how valid it is to run any part of it as an independent executable. If it works, it's probably ok, but I'd definitely check with the infra team.

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: > Justin alerted me that we have some exec helpers in build system I'll check out.

Ah, can't use these just yet. Need to update the exec helpers to support the args concept, since each platform needs different args escaping.

**Comment by samouri** [comment]
@ampproject/wg-infra: PTAL and let me know if you have concerns about the approach. I tried using the `exec` helpers, but it will require refactoring them to accept args. For simple cases like the git integrations its fine to simply throw in args in the cmd, but for complex ones likes this that require shell-dependent args escaping, we actually need to take advantage of the full nodejs api.

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js [hunk 3]
```diff
@@ -126,3 +127,21 @@ module.exports = function ({types: t, template}) {
...
+
+  // TODO: migrate to the helpers in build-system exec.js
+  // after adding args support.
+  const spawnedProcess = spawnSync('node', ['-e', programText], {
+    cwd: __dirname,
+    env: process.env,
+    encoding: 'utf-8',
+    stdio: 'pipe',
+  });
+  return spawnedProcess.stdout;
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-jss/index.js:146]: We should be checking the return code and `stderr` to see if any errors happened.

File: build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-var _classes = JSON.parse("{\"button\":\"button-21aa4a8\",\"CSS\":\".button-21aa4a8 {\\n  font-size: 12px;\\n}\"}");
+var _classes = JSON.parse("{\"button\":\"button-21aa4a8\",\"CSS\":\".button-21aa4a8{font-size:12px}\\n\"}");
```
Comment (jridgewell) [inline build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs:1]: I'm surprised the trailing `\\n` is still here.

**Review by jridgewell — APPROVED** [review]

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js (hunk 3 above)
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:146]: Good thinking. If anything is written to `stderr` or if there is a bad exit code we'll want to throw.

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:146]: > We should be checking the return code and stderr to see if any errors happened.

Ok, so we actually shouldn't use `stderr` to determine failure, since even warnings go there.

File: build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs (hunk 4 above)
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs:1]: 🤷

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js (hunk 3 above)
Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:151]: Yah, this is the correct way to do it.

File: build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs (hunk 4 above)
Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs:1]: And we add a new test case that has multiple classes? I wan to see if whitespace between rules is stripped.

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs:1]: Can modify an existing test case with a second ruleset

Comment (jridgewell) [reply build-system/babel-plugins/babel-plugin-transform-jss/test/fixtures/transform-assertions/should-transform-jss-var/output.mjs:1]: Sure

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js (hunk 2 above)
Comment (rcebulko) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: When we spawn a process like this, don't we get access to the stdin/stdout file handles? My hunch is maybe this could be a standalone script that reads from stdin and transforms that CSS. That in turn could play nicer with gulp/parallelization, and completely avoid the escaping-command-line-arguments issue. WDYT(?

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: It could be! I'm not sure if that helps though, since it would only cause more latency for the gulp css task, and this already solves the encoding issue for the babel transform.

**Comment by samouri** [comment]
next up: enumification

File: build-system/babel-plugins/babel-plugin-transform-jss/index.js (hunk 2 above)
Comment (rsimha) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: Piling on to this thread with some alternative ideas. First off, do I understand correctly that if `transformCss()` from `jsify-css.js` were synchronous, we wouldn't need the workaround in this file? And is it true that the thing that makes `transformCss()` asynchronous is this call to `postcss().process()`?

https://github.com/ampproject/amphtml/blob/32579ce377e0849f04966951fc3123aef8f751d9/build-system/tasks/jsify-css.js#L75-L77

Here are two alternatives that come to mind:
1. From looking at this code, can we use `postcss-js`, which appears to have a [synchronous API](https://github.com/postcss/postcss-js#syncplugins-function)?
2. Modify the code in `jsify-css.js` to work [directly on the `gulp` stream](https://www.npmjs.com/package/postcss#gulp), and wrap it to create a synchronous function that can be called here.

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-jss/index.js:133]: Yep, if it were sync then we wouldn't need to call `spawnSync` here, and `postcss` is what causes async.

1. I tried the sync api, but it throws since all of our plugins can only operate in async mode :/
2. I don't know enough about `gulp` to say if this makes sense, but if the streams abstraction lets us easily turn this async function sync (via abstractions over subprocesses), then that would definitely be better! Only better for the babel transforms though, the previous callers were fine async
